Homeworks/Homework SQL/SQLQuery-Homework.sql

```diff
@@ -0,0 +1,339 @@
+FROM Librarians l
+INNER JOIN Libraries i ON i.Id = l.LibraryId
+ORDER BY l.Name ASC,i.Name DESC;
+
+--3. Update all the street names to start with "STR.".
+
+UPDATE LIbraries
+SET
+    StreetName = 'STR.'+StreetName
+WHERE Id=11;
+
+SELECT *
+FROM Libraries;
+
+--4. Display the authors' email for those who have more than 3 books published.
+
+SELECT a.Email
+FROM Author a 
+RIGHT JOIN Books b ON b.AuthorId=a.Id
+GROUP BY  a.Email, a.Id
+HAVING COUNT(a.Id) >3;
+
+--5. Display the library name and the book title of the largest sale (the highest number of copies sold in one sale).
+
+SELECT l.Name, bb.Title
+FROM Libraries l 
+JOIN Sales s ON s.LibraryId= l.Id 
+JOIN Books_Sales b ON  b.SalesId= s.Id
+JOIN Books bb ON bb.Id= b.BookId
+WHERE s.NumberOfCopies = ( SELECT MAX (NumberOfCopies) FROM Sales);
+
+--6. Display all the librarians that are not on holiday and are hired more than 5 years ago.
+
+SELECT l.Name, i.HolidayStatus as IsOnHoliday
+FROM Librarians l
+JOIN IsOnHoliday i ON i.Id=l.IsOnHolidayId
+WHERE i.Id=2 AND l.HireDate< DATEADD(year,-5,GETDATE());
+
+--7. Create a stored procedure that displays, based on the book title, the number of libraries in which the book is found. Call the procedure with three different book titles.
+
+GO
+CREATE PROCEDURE nrOfLibraries @BookTitle varchar(20)
+AS
+BEGIN
+    SELECT b.Title, COUNT(l.LibraryId) as NumberOfLibraries
+    FROM   Books b
+    JOIN   Books_Libraries l ON l.BookId= b.Id
+    WHERE b.Title= @BookTitle
+    GROUP BY b.Title
+END
+GO
+
+EXEC nrOfLibraries @BookTitle='Book1';
+EXEC nrOfLibraries @BookTitle='Book2';
+EXEC nrOfLibraries @BookTitle='Book3';
+
+--8. Create a function that returns, based on the author's email, the number of copies sold for each of his / her books. Call the function with three different emails.
+GO
+CREATE FUNCTION dbo.NrOfCopies(@AuthorMail varchar(30)) 
+RETURNS int
+AS
+BEGIN
+    DECLARE @Copies INT
+    SET @Copies =
+    (SELECT s.NumberOfCopies 
+    FROM Sales s
+    JOIN Books_Sales bs ON bs.SalesId=s.Id
+    JOIN Books b ON b.Id= bs.BookId
+    JOIN Author a ON a.Id= b.AuthorId
+    WHERE a.Email= @AuthorMail
+    GROUP BY s.NumberOfCopies)
+    RETURN @Copies;
+END 
+GO
+
+SELECT dbo.NrOfCopies('am@gmail.com');
+
+--9. Create a trigger so that when a new librarian is added to the system with no library assigned (NULL on LibraryId), he / she is automatically assigned to a library.
+
+Go
+CREATE TRIGGER AddStandardLibrarian
+ON Librarians
+AFTER INSERT
+AS
+BEGIN
+    UPDATE Librarians
+    SET LibraryId= 1
+    WHERE Librarians.LibraryId=NULL
+END 
+GO
+
+--10. Create a transaction that should have at least three commands inside it. Based on the logic you think of use at least one savepoint and at least one rollback operation.
+
+BEGIN TRAN 
+INSERT INTO Librarians
+VALUES ('Giovanni','2022-01-01',1,2);
+ROLLBACK TRAN
+UPDATE LIbraries
+SET
+    Name= 'Carturesti'
+WHERE Id=1;
+SAVE TRANSACTION AddLibrarian
```

Comment:
There is an error that says "Cannot issue SAVE TRANSACTION when there is no active transaction." on this line. This happens because there is the rollback tran command which ends the transaction and a new one is not started.